docs/design/2024-09-06-maintain-priority-queue-in-memory.md

```diff
@@ -0,0 +1,567 @@
+}
+
+func (n *DDLNotifier) deleteProcessedEvent(sctx sessionctx.Context, eventID int) error {
+    _, err := ExecRows(sctx, "DELETE FROM mysql.ddl_notifier WHERE job_id = ?", eventID)
+    return err
+}
+
+func (n *DDLNotifier) RegisterHandler(id int, handler SchemaChangeHandler) {
+    n.handlers[id] = handler
+}
+```
+
+This notifier is responsible for fetching events from the `mysql.ddl_notifier` table, ordered by job ID. After it calls every handler for each event, it can safely delete handled events.
+We use a separate TXN to handle each handler. This would help us reduce the impact between each handler. After all handlers have handled this event, we will delete it.
+
+##### Where does it run?
+
+Since only the stats module is currently subscribed and requires in-memory operations, we will only start it at the stats owner.
+
+##### Future possibility
+
+If we do want other modules to subscribe to the module in the future, we need to avoid it only running in the stats owner node. This is because other modules may also have memory operations that may not occur in the stats owner node.
+**A possible solution would be to start DDLNotifier on each TiDB node, with different subscribers on each node. This would allow us to solve the problem.**
+However, this solution introduces more complexities, such as how to manage subscribers, and ensuring that an event can be deleted after it has been processed by subscribers on all the different nodes becomes more complicated. Even the granularity of the subscribers needs to be controlled. For example, when the same subscriber logic is used, but they are running on every node, whether we consider it to be a single subscriber or multiple subscribers is something that needs to be discussed.
+**Since this issue is already far from our goal this time, I decided to consider it as a future possibility rather than in the pursuit of completing it in this design.**
+
+#### Reflect on Time Change
+
+We need to consider a time factor here because one of the metrics used to calculate our scores is the analysis interval, which is the time interval between the last analysis and the present. Obviously, it changes over time and we need to keep recalculating the analysis interval for these tables.
+We can recalculate the last analysis intervals and scores for these tables every 5 minutes using the background worker behind the priority queue.
+
+#### Failover
+
+Currently, there are only two scenarios where failover needs to be handled.
+
+1. the node is restarted and
+2. an owner switch occurs.
+
+In these two cases, we need to rebuild the priority queue, that is, to rescan and reconstruct the entire priority queue.
+For DML events, we scan the entire stats cache for the latest `modify_count` and `count`.
+For DDL events, since all our operations are processed in transactions. So after waiting for the priority queue to be ready, we just let the notifier restart processing.
```

Comment:
I only understand that the transaction consumes the mysql.ddl_notifier` table, but where is the table that the results are saved? Do we persist the enrties of heap in a table?

Reply:
No, we just rebuilding it with the latest schema and the stats cache.

Reply:
Offline discussion:
Because pq is time-consuming in the rebuild process, its api returns errors that can be retried until it finishes rebuilding, and then the notifier holds those events. Waiting until PQ rebuilds before sending them to it is acceptable for PQ to receive a duplicate ddl event since we always go with the latest schema.